Iter9

Tati-till opened the pull request.

Tati-till committed:
shortener: Fix handling of incorrect URLs and add a test

Tati-till committed:
shortener: Add logger

Tati-till committed:
shortener: Add a new endpoint to handle URLs in JSON format

Tati-till committed:
shortener: Add gzip support

Tati-till committed:
shortener: Save all stored URLs to a file and load them on service startup

nburunova commented on cmd/shortener/gzip.go:
```diff
@@ -0,0 +1,110 @@
+package main
```
этот файл лучше переместить в папку middleware, в следующих спринтах будут добавляться еще middleware и спрятать ее в internal

Tati-till replied:
Fixed.

nburunova commented on cmd/shortener/handlers.go:
```diff
@@ -1,15 +1,69 @@
 package main
 
 import (
+	"encoding/json"
```
хендлеры тоже лучше переместить в папку internal/handlers. Будут добавляться еще хендлеры в будущем

Tati-till replied:
Fixed.

nburunova commented on cmd/shortener/handlers.go:
```diff
@@ -1,15 +1,69 @@
+	var receivedReq models.Request
+	err = json.Unmarshal(received, &receivedReq)
+	if err != nil {
+		http.Error(res, "Can't read body", http.StatusBadRequest)
```
тут советую изменить текст ошибки, потом ты не поймешь где произошла ошибка: в 16 строке или в 23, так как в тексте увидишь Can't read body для обоих случаев

Tati-till replied:
Fixed.

nburunova commented on cmd/shortener/handlers.go:
```diff
@@ -1,15 +1,69 @@
+
+	hash, err := generator(receivedReq.URL)
+	if err != nil {
+		http.Error(res, err.Error(), http.StatusInternalServerError)
```
тут ты еще возвращаешь текст ошибки err.Error(), так же можно поступать в случае ошибок в выше , в 18 и 25 строках. Будет информативнее

Tati-till replied:
Fixed.

nburunova commented on cmd/shortener/handlers.go:
```diff
@@ -1,15 +1,69 @@
+
+	_, err = res.Write(resJSON)
+	if err != nil {
+		fmt.Println("Failed to write response:", err)
```
возможно стоит логировать ошибку через логгер. fmt для логирования редко применяется

Tati-till replied:
Fixed.

nburunova commented on cmd/shortener/main.go:
```diff
@@ -4,31 +4,80 @@ import (
+	initStorage(conf)
+
+	// Close producer and consumer.
+	defer func(Producer *file.Producer) {
```
названия аргументов ф-ций пишем обычно с маленькой буквы

Tati-till replied:
Fixed.

nburunova commented on cmd/shortener/main.go:
```diff
@@ -4,31 +4,80 @@ import (
-	var err error
-	storageURLs, err = store.NewStore()
+var (
+	storageURLs store.Store
```
старайся не применять глобальные переменные. Это может быть источником проблем. Код будет неочевиден. Если функции нужен объект storage и она собирается его менять \ читать, то передай ей storage явно в качестве аргумента. Тогда при чтении кода будет сразу ясно: ага, эта фукнция будет что-то делать со storage. Если будет глобальная переменная, то функции могут обращаться к storage неявно, сложно будет отлаживаться и контролировать , что происходит со storage.   
Лучше создать storage  в main, storage пусть создает и возвращает функция createStorage, переделай ее из функции initStorage.

То же самое для переменных producer, consumer - создавай их в main и явно передавай их в фукнции в качестве аргументов, если producer\consumer там нужны.

nburunova replied:
это еще и интерфейс. Если ты объявила глобальную переменную типа интерфейса в main значит чтото не так. Обычно переменные типа интерфейс передаются в конструукторы других объектов. И интерфейсы объявляются там, в тех пакетах, кому будет нужен store . И методы в интерфейсе объявляются именно те, что нужны этому пакету.

nburunova replied:
Почитай https://habr.com/ru/articles/856272/ - раздел Где размещать интерфейсы

Tati-till replied:
Изменила, вроде работает. Передаю как ресивер, потому что не поняла из всего, что написано выше как передать storage как параметр. Назначила 1-2-1, надеюсь, объясните.

nburunova commented on internal/storage/file/store.go:
```diff
@@ -0,0 +1,80 @@
+	return p.file.Close()
+}
+
+func (p *Producer) WriteRecord(record *Record) error {
```
проверь, всё ли будет в порядке, если несколько горутин будут одновременно вызывать WriterRecord. Не будет ли гонки? Все ли записи попадут в файл?

Tati-till replied:
добавила Mutex

ekubyshin replied:
мы еще не проходили тему синхронизации и горутин, поэтому на это пока можно не обращать внимания. Но в целом в го несколкьо подходов для синхронизации, мьютекс - один из них.

nburunova commented on internal/storage/store.go:
```diff
@@ -1,14 +1,16 @@
 	"shorturl/internal/storage/memory"
 )
 
 type Store interface {
```
интерфейс объявляется там, где он используется. 
https://go.dev/wiki/CodeReviewComments#interfaces 
Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.

nburunova commented on internal/storage/store.go:
```diff
@@ -1,14 +1,16 @@
 
-func NewStore() (Store, error) {
-	return memory.NewStore(), nil
+func NewStore(consumer *file.Consumer, producer *file.Producer) (Store, error) {
```
этот конструктор не нужен. Сразу используй memory.NewStore(consumer, producer).

В main сразу вызывай memory.NewStore(consumer, producer), а потом передавай этот объект туда, где испльзуется store

Tati-till replied:
Fixed.

Tati-till committed:
handlers: Refactoring returned errors

Tati-till committed:
*: Refactoring

- Moved `gzip.go` to `middleware/`.
- Moved `handlers.go` to `handlers/`.
- Reorganized the `Storage` interface and moved it to `handler.go`.
- Removed the global storage variable in `main.go`.
- Used `Storage` as a receiver for handler methods.

Tati-till committed:
file/store: Add Mutex to `Producer` struct to prevent race conditions

ekubyshin commented on cmd/shortener/handlers_test.go:
```diff
@@ -162,10 +182,85 @@ func Test_mainHandler(t *testing.T) {
 
 	config.ParseFlags()
-	ts := httptest.NewServer(mainRouter())
+	conf := config.GetConfig()
```
NewConfig - конструтор в го принято называть так.

Tati-till replied:
@ekubyshin У меня это не совсем конструктор, я после парсинга флагов получаю конфигурационные параметры. Все равно стоит переименовать в этом случае? https://github.com/Tati-till/shorturl/blob/e7c27b03b83be237a782e8eed5019aaccacdce89/internal/config/config.go#L33-L35

ekubyshin commented on cmd/shortener/handlers_test.go:
```diff
@@ -162,10 +182,85 @@ func Test_mainHandler(t *testing.T) {
-	ts := httptest.NewServer(mainRouter())
+	conf := config.GetConfig()
+
+	producer, err := file.NewProducer(conf.FileStoragePath)
```
тут конструкторы все названы в едином стиле. Единый стиль именования, написания кода - важный момент при разработке проекта, тк облегчает написание кода несколькими разработчиками и проведения ревью.

ekubyshin commented on cmd/shortener/main.go:
```diff
@@ -43,22 +80,14 @@ func mainRouter() chi.Router {
 
 	r.Route("/", func(r chi.Router) {
-		r.Post("/", generateURL) // POST /
+		r.Post("/", logger.WithLogging(middleware.GzipMiddleware(h.GenerateURL))) // POST /
```
миддлвары подключаются через r.Use(middleware.Logger)

ekubyshin commented on internal/config/config.go:
```diff
@@ -6,15 +6,17 @@ import (
 type Config struct {
-	RunAddr string
-	ResAddr string
+	RunAddr         string
```
В теме Os изучали парсинг конфиги через различные библиотеки. Предлагаю использовать готовую, тк в реальных проектах переменных окружения довольно много и парсить каждую вручную будет довольно утомительным занятием. 
https://github.com/caarlos0/env - здесь есть возможность указывать дефолтовые значения, если оно отсутствует в env.

ekubyshin commented on internal/config/config.go:
```diff
@@ -6,15 +6,17 @@ import (
 
 var c Config
 
 func ParseFlags() {
```
NewConfig - который конструирует весь конфиг и возвращает готовый к использованию объект.
NewConfig() (*Config, error){} - конструктор конфига

ekubyshin commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,149 @@
+	"shorturl/internal/models"
+)
+
+type Storage interface {
```
тут отлично, что по го стилю

ekubyshin commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,149 @@
+		return
+	}
+
+	if receivedReq.URL == "" {
```
для валидации URL выгоднее воспользоваться методом ParseURL из пакета http

ekubyshin commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,149 @@
+	return err == nil
+}
+
+func getHashFromURL(url []byte) string {
```
это на мой взгляд лучше вынести в отдельный пакет hash и положить как отдельный метод, который можно использовать там, где он нужен.

ekubyshin commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,149 @@
+	res.WriteHeader(http.StatusTemporaryRedirect)
+}
+
+func isCorrectURL(s string) bool {
```
также вынести в пакет validation - чтобы было больше декомпозиции ответственности

ekubyshin commented on internal/logger/logger.go:
```diff
@@ -0,0 +1,81 @@
+	return nil
+}
+
+func WithLogging(h http.HandlerFunc) http.HandlerFunc {
```
Предлагаю сделать декомпозицию ответственности и вынести в пакет middleware

Tati-till committed:
handler: Check URL in JSON

Tati-till committed:
hash: Move function `GetHashFromURL` to a separate `hash` package

Tati-till committed:
validation: Move function `IsCorrectURL` to a separate `validation` package

Tati-till committed:
middleware: Move `WithLogging` to `middleware` package
